paddle/cinn/ir/ir_schedule_error.h

```diff
@@ -0,0 +1,173 @@
+#include <execinfo.h>
+#endif
+
+#include <iostream>
+#include <memory>
+#include <sstream>
+#include <stdexcept>
+#include <string>
+#include <vector>
+
+namespace cinn {
+namespace ir {
+
+namespace enforce {
+
+#ifdef __GNUC__
+inline std::string demangle(std::string name) {
+  int status = -4;  // some arbitrary value to eliminate the compiler warning
+  std::unique_ptr<char, void (*)(void*)> res{
+      abi::__cxa_demangle(name.c_str(), NULL, NULL, &status), std::free};
+  return (status == 0) ? res.get() : name;
+}
+#else
+inline std::string demangle(std::string name) { return name; }
+#endif
+
+static std::string GetErrorSumaryString(const std::string& what,
+                                        const char* file,
+                                        int line) {
+  std::ostringstream sout;
+  sout << "\n----------------------\nError Message "
+          "Summary:\n----------------------\n";
+  sout << what << "(at " << file << " : " << line << ")" << std::endl;
+  return sout.str();
+}
+
+static std::string GetCurrentTraceBackString() {
+  std::ostringstream sout;
+  sout << "\n\n--------------------------------------\n";
+  sout << "C++ Traceback (most recent call last):";
+  sout << "\n--------------------------------------\n";
+#if !defined(_WIN32) && !defined(PADDLE_WITH_MUSL)
+  static constexpr int TRACE_STACK_LIMIT = 100;
+
+  void* call_stack[TRACE_STACK_LIMIT];
+  auto size = backtrace(call_stack, TRACE_STACK_LIMIT);
+  auto symbols = backtrace_symbols(call_stack, size);
+  Dl_info info;
+  int idx = 0;
+  int end_idx = 0;
+  for (int i = size - 1; i >= end_idx; --i) {
+    if (dladdr(call_stack[i], &info) && info.dli_sname) {
+      auto demangled = demangle(info.dli_sname);
+      std::string path(info.dli_fname);
+      // C++ traceback info are from core.so
+      if (path.substr(path.length() - 3).compare(".so") == 0) {
+        sout << idx++ << " " << demangled << "\n";
+      }
+    }
+  }
+  free(symbols);
+#else
+  sout << "Not support stack backtrace yet.\n";
+#endif
+  return sout.str();
+}
+
+static std::string GetTraceBackString(const std::string& what,
+                                      const char* file,
+                                      int line) {
+  return GetCurrentTraceBackString() + GetErrorSumaryString(what, file, line);
+  // return GetErrorSumaryString(what, file, line);
+}
+
+struct EnforceNotMet : public std::exception {
+ public:
+  EnforceNotMet(const std::string& str, const char* file, int line)
+      : err_str_(GetTraceBackString(str, file, line)) {}
+
+  const char* what() const noexcept override { return err_str_.c_str(); }
+
+ private:
+  std::string err_str_;
+};
+
+#define CINN_THROW(...)                                              \
+  do {                                                               \
+    try {                                                            \
+      throw enforce::EnforceNotMet(__VA_ARGS__, __FILE__, __LINE__); \
```

Comment:
Is this same as `Paddle` enforce? As I remember, we discussed offline that we use `PADDLE_THROW` when we found the definition, otherwise we define by ourself, then we can handle it with both CINN-only and Paddle-CINN.

Does this implementation also handle both CINN-only and Paddle-CINN? Do we reuse Paddle code?

Reply:
`PADDLE_THROW` is more complicated and has many redundant functions. If we want to reuse Paddle code we have to include more Paddle header files, thus we cannot build CINN-ONLY.

Reply:
OK, not best but fine to me now.